### DISCUSSION TITLE
Nested Subcommands: Context sensitive argument parsing

### DISCUSSION THREAD

**Discussion by 7tupel** [babashka/cli#62] [Q&A]
Hello everyone,

I stumbled into the following situation: I'm trying to build a simple command line interface to run jobs that accepts commands in the form of
```sh
run job <job> <options>
```

Thus far this is easy to implement using the Subcommand mechanism build in bb-cli i.e.
```clojure
(cli/dispatch
[{:cmds ["run" "job"]
   :fn run-job
   :args->opts [:job]}])
```

In my usecase, every job has its own unique set of additional options. I.e. _jobA_ takes the option `--path "/some/path"` and _jobB_ takes `--depth 2`. 

What I'm looking for is a way to delegate the job specific options down to the function that runs the specific job and use the cli spec and coercion mechanism there to parse and coerce the job's options. I.e. something like this:
```clojure
(defn jobA
  [args]
  (let [options  (cli/parse-opts args {:spec {:paths {:coerce :string :default "foo/bar"}}})]
  ;...
))
```

Of cause the `args` that are passed down to `jobA` by the `dispatch` function above are a map with the command line arguments already parsed. Since the function that contains the dispatch does not now the spec down in jobA, the coercion might be wrong.

Is there a build in mechanism I can pass down the original `*command-line-args*` and parse them in the job specific function? Or another mechanism to do what I would like to achieve? (except from using `*command-line-args*` inside the job function)

**Comment by borkdude** [comment]
You can do this with:

``` clojure
(cli/dispatch
[{:cmds ["run" "job"]
   :fn run-job
   :spec {:paths {:coerce :string :default "foo/bar"}}
   :args->opts [:job]}]  ...args)
```

All options in a subcommand entry are merged.

**Reply by 7tupel** [reply]
But this will only work if all jobs use the same option format. I.e. when jobA expects a `paths` option that is a string with comma separated paths and another jobC expects a `paths` option as a vector of strings that each are paths. In that case merging would not simply work is that correct?

**Reply by borkdude** [reply]
@7tupel In that case each job becomes its own subcommand:

```
{:cmds ["run" "job" "jobA"]
 :spec ...}
{:cmds ["run" "job"] ;; fallback for other jobs
 :spec ...}
```

**Reply by borkdude** [reply]
Alternatively you can first parse the cmds and then do stuff with it after the fact:

```
user=> (cli/parse-cmds ["foo" "bar" "-baz"])
{:cmds ("foo" "bar"), :args ("-baz")}
```

**Reply by 7tupel** [reply]
Thanks a lot. I was hoping to be able to avoid creating a single command for each job since users can add their own custom jobs. This way I need to build the vector for the dispatch function dynamically which I was trying to avoid.

**Reply by borkdude** [reply]
We could pass the raw args to the job function by adding a key here:

https://github.com/babashka/cli/blob/d011ad3516be08c829a98013b3839d10b5e4fef2/src/babashka/cli.cljc#L583

if that would help?

**Reply by 7tupel** [reply]
maybe this would work. I'll have to try
